sim/midas/src/main/cc/core/widget_registry.h

```diff
@@ -0,0 +1,119 @@
+// See LICENSE for license details.
+
+#ifndef __BRIDGE_REGISTRY_H
+#define __BRIDGE_REGISTRY_H
+
+#include <cassert>
+
+#include <memory>
+#include <unordered_map>
+#include <vector>
+
+#include "core/widget.h"
+
+class FASEDMemoryTimingModel;
+class bridge_driver_t;
+class widget_t;
+class StreamEngine;
+
+/**
+ * Unique class responsible for the creating and ownership of all bridges.
+ *
+ * The bridges of a design are registered with the registry through the
+ * `add_widget` methods for them to be retrieved later through the appropriate
+ * getters.
+ */
+class widget_registry_t final {
+public:
+  widget_registry_t(const TargetConfig &config,
+                    simif_t &simif,
+                    const std::vector<std::string> &args);
+
+  ~widget_registry_t();
+
+  /**
+   * Return a list of pointers to all bridges of a type.
+   *
+   * @tparam T Type of bridge to fetch.
+   * @return List of non-owning pointers to bridges.
+   */
+  template <typename T>
+  std::vector<T *> get_bridges() {
+    std::vector<T *> bridge_list;
+    for (auto &bridge : widgets[&T::KIND]) {
+      bridge_list.push_back(static_cast<T *>(bridge.get()));
+    }
+    return bridge_list;
+  }
+
+  /**
+   * Return a widget of a particular kind which has a single instance.
+   *
+   * This function should only be used with widgets that have a unique instance.
+   * If multiple widgets of the same kind were registered or the widget does
+   * not exist, the function fails.
+   *
+   * @tparam T Type of the widget to fetch
+   * @return Reference to the widget.
+   */
+  template <typename T>
+  T &get_widget() {
```

Comment:
This is more of a musing but i'd probably prefer an `std::optional` flavor of this function as well (or instead of).

Reply:
A pointer version will come up in a subsequent PR.